File: src/dashboard/apigateway/apigateway/controller/helm/chart.py
#
# TencentBlueKing is pleased to support the open source community by making
# 蓝鲸智云 - API 网关(BlueKing - APIGateway) available.
# Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
# Licensed under the MIT License (the "License"); you may not use this file except
# in compliance with the License. You may obtain a copy of the License at
#
#     http://opensource.org/licenses/MIT
#
# Unless required by applicable law or agreed to in writing, software distributed under
# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
# either express or implied. See the License for the specific language governing permissions and
# limitations under the License.
#
# We undertake not to change the open source license (MIT license) applicable
# to the current version of the project delivered to anyone in the future.
#
import logging
from typing import Optional
from urllib.parse import urlparse

from attrs import define
from bkapi.bcs_api_gateway.client import Client as BcsApiGatewayClient
from django.conf import settings

from apigateway.components.bcs import get_bcs_api_gateway_client
from apigateway.components.bcs_helper import BcsApiGatewayApiRequestError
from apigateway.components.chart import RepoClient
from apigateway.controller.exceptions import ChartRepoNotFound, ChartUploadFailed
from apigateway.utils.exception import check_result_code

logger = logging.getLogger(__name__)


@define(slots=False)
class ChartRepoInfo:
    project_name: str
    name: str
    type: str
    url: str
    username: str
    password: str


@define(slots=False)
class ChartInfo:
    name: str
    version: str


@define(slots=False)
class ChartHelper:
    client: Optional[BcsApiGatewayClient] = None
    access_token: str = ""

    def __attrs_post_init__(self):
        if not self.client:
            self.client = get_bcs_api_gateway_client()
        if self.access_token and self.access_token != "":
            self.client.update_bkapi_authorization(access_token=self.access_token)

    def get_public_repo_info(self) -> ChartRepoInfo:
        """Get the public chart repos info."""
        return self.get_repo_info(settings.BCS_PUBLIC_CHART_PROJECT, settings.BCS_PUBLIC_CHART_REPOSITORY)

    def get_project_repo_info(self, project_name: str) -> ChartRepoInfo:
        """Get the project chart repos info."""
        return self.get_repo_info(project_name, project_name)

    def get_repo_info(self, project_name: str, repo_name: str) -> ChartRepoInfo:
        """Get the chart repo info by name."""
        assert self.client
        result = self.client.api.get_repo(
            path_params={
                "projectCode": project_name,
                "name": repo_name,
            }
        )

        check_result_code("query chart repo", BcsApiGatewayApiRequestError, result.get("code"), result.get("message"))

        if not result.get("result"):
            raise ChartRepoNotFound(repo_name)

        repo = result["data"]
        return ChartRepoInfo(
            project_name=repo["projectCode"],
            name=repo["name"],
            type=repo["type"],
            url=repo["repoURL"],
            username=repo["username"],
            password=repo["password"],
        )

    def push_chart(self, chart_file: str, repo_info: ChartRepoInfo):
        """Push the chart to the repo."""

        url = urlparse(repo_info.url)
        repo_info = self.get_repo_info(repo_info.project_name, repo_info.name)
        client = RepoClient(
            endpoint=f"{url.scheme}://{url.netloc}",
            project_name=repo_info.project_name,
            repo_name=repo_info.name,
        )
        client.set_auth(repo_info.username, repo_info.password)

        upload_result = client.api.upload(files={"chart": open(chart_file, "rb")}, params={"force": "1"})
        if not upload_result.get("saved"):
            raise ChartUploadFailed()

    def check_chart_exists(self, project_name: str, repo_name: str, name: str, version: str, operator: str):
        """Check the chart exists in the repo."""
        assert self.client

        result = self.client.api.get_chart_versions(
            path_params={
                "projectCode": project_name,
                "repoName": repo_name,
                "name": name,
            },
            params={"operator": operator},
        )

        check_result_code(
            "query chart versions", BcsApiGatewayApiRequestError, result.get("code"), result.get("message")
        )

        if not result["result"]:
            return False

        data = result["data"]
        for i in data["data"]:
            if i.get("version") == version:
                return True

        return False

File: src/dashboard-front/src/mixins/chart.js
/*
 * TencentBlueKing is pleased to support the open source community by making
 * 蓝鲸智云 - API 网关(BlueKing - APIGateway) available.
 * Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 *     http://opensource.org/licenses/MIT
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * We undertake not to change the open source license (MIT license) applicable
 * to the current version of the project delivered to anyone in the future.
 */
import dayjs from 'dayjs'

export default {
  methods: {
    $getChartIntervalOption (max, type, axis) {
      const option = {
        yAxis: {
          axisLabel: {}
        },
        xAxis: {
          axisLabel: {}
        }
      }

      // 动态计算出max合适值和interval配置
      if (type === 'number') {
        if (max >= 0 && max <= 10) {
          option[axis].interval = 1
        } else if (max > 10 && max <= 20) {
          option[axis].interval = 2
        } else if (max > 20 && max <= 50) {
          option[axis].interval = 5
        } else if (max > 50 && max <= 100) {
          option[axis].interval = 10
        } else if (max > 100 && max <= 1000) {
          option[axis].interval = 100
        } else if (max > 1000 && max <= 5000) {
          option[axis].interval = 500
        } else if (max > 5000 && max <= 10000) {
          option[axis].interval = 1000
        } else if (max > 10000 && max <= 20000) {
          option[axis].interval = 2000
        } else if (max > 20000 && max <= 50000) {
          option[axis].interval = 5000
        } else if (max > 50000 && max <= 100000) {
          option[axis].interval = 10000
        } else if (max > 100000 && max <= 500000) {
          option[axis].interval = 20000
        } else if (max > 500000 && max <= 1000000) {
          option[axis].interval = 50000
        } else if (max > 1000000 && max <= 5000000) {
          option[axis].interval = 500000
        } else if (max > 5000000 && max <= 10000000) {
          option[axis].interval = 1000000
        } else {
          option[axis].interval = 10000000
        }
        option[axis].max = function (value) {
          return value.max + (option[axis].interval - (value.max % option[axis].interval))
        }

        option[axis].min = function (value) {
          // 使0值显示在grid中间
          if (value.max === 0) {
            return -1 * (value.max + (option[axis].interval - (value.max % option[axis].interval)))
          }

          return 0
        }

        // option[axis].axisLabel.formatter = (value) => {
        //     let displayValue = value
        //     if (value >= 1e3) {
        //         displayValue = `${value / 1e3} K`
        //     }
        //     return displayValue
        // }
      }

      // 根据时间值计算显示年/月/日/时间部分
      if (type === 'time') {
        if (max >= 0 && max <= 3600) {
          // 1小时内
          option[axis].interval = 60 * 1000 * 5
          option[axis].axisLabel.formatter = (value) => {
            return dayjs(value).format('HH:mm:ss')
          }
        } else if (max > 3600 && max <= 86400) {
          // 1天内
          option[axis].interval = 60 * 60 * 1000 * 2
          option[axis].axisLabel.formatter = (value) => {
            return dayjs(value).format('HH:mm')
          }
        } else if (max > 86400 && max <= 86400 * 31) {
          // 1月内
          option[axis].interval = 86400 * 1000
          option[axis].axisLabel.formatter = (value) => {
            return dayjs(value).format('MM-DD HH:mm')
          }
        } else if (max > 86400 * 31 && max <= 86400 * 31 * 12) {
          // 1年内
          option[axis].interval = 86400 * 1000 * 31
          option[axis].axisLabel.formatter = (value) => {
            return dayjs(value).format('MM-DD')
          }
        } else {
          // 大于1年
          option[axis].interval = 86400 * 1000 * 31 * 3
          option[axis].axisLabel.formatter = (value) => {
            return dayjs(value).format('YYYY-MM-DD')
          }
        }
      }

      return option
    }
  }
}
